pkg/registry/comparison/comparison.go

```diff
@@ -0,0 +1,178 @@
+package comparison
+
+import (
+	"context"
+	"encoding/csv"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/caicloud/nirvana/log"
+	"github.com/kleveross/ormb/pkg/model"
+	ormbmodel "github.com/kleveross/ormb/pkg/model"
+
+	"github.com/kleveross/klever-model-registry/pkg/common"
+	"github.com/kleveross/klever-model-registry/pkg/registry/errors"
+	"github.com/kleveross/klever-model-registry/pkg/registry/harbor"
+	"github.com/kleveross/klever-model-registry/pkg/registry/paging"
+	"github.com/kleveross/klever-model-registry/pkg/util"
+)
+
+// Generator list models' metadata and compare
+func Generator(ctx context.Context, models Comparison, opt *paging.ListOption) (*paging.Result, error) {
+	metaList, err := composeComparison(models.Models)
+	if err != nil {
+		return nil, err
+	}
+	return paging.Page(metaList, opt), nil
+}
+
+func composeComparison(models []ComparisonModel) ([]model.Model, error) {
+	proxy := harbor.NewProxy(common.ORMBDomain, common.ORMBUserName, common.ORMBPassword)
+	metaList := make([]ormbmodel.Model, 0)
+	for _, model := range models {
+		artifacts, err := proxy.ListArtifacts(model.Project, model.Name)
+		if err != nil {
+			return nil, err
+		}
+		for _, artifact := range artifacts {
+			for _, tag := range artifact.Tags {
+				if tag.Name == model.Tag {
+					manifest, err := json.Marshal(artifact.ExtraAttrs)
+					if err != nil {
+						return nil, err
+					}
+					var meta ormbmodel.Metadata
+					if err := json.Unmarshal(manifest, &meta); err != nil {
+						return nil, err
+					}
+					metaList = append(metaList, ormbmodel.Model{
+						Path:     fmt.Sprintf("%s/%s:%s", model.Project, model.Name, model.Tag),
+						Metadata: &meta,
+					})
+					break
+				}
+			}
+		}
+	}
+
+	return metaList, nil
+}
+
+// DownloadCSVFile downloads the comparison csv file
+func DownloadCSVFile(ctx context.Context, models Comparison) error {
+	metaList, err := composeComparison(models.Models)
+	if err != nil {
+		return err
+	}
+	err = createCSVFile(ctx, metaList)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func createCSVFile(ctx context.Context, metas []ormbmodel.Model) error {
+	fileName := fmt.Sprintf("%d.csv", time.Now().Unix())
+
+	file, err := os.Create(fileName)
+	if err != nil {
+		return fmt.Errorf("open file is failed, err: %s", err.Error())
+	}
+
+	defer func() {
+		if err := file.Close(); err != nil {
+			log.Error("close file is failed, err: %s", err.Error())
+		}
+		if err := os.Remove(fileName); err != nil {
+			log.Error("remove file is failed, err: %s", err.Error())
+		}
+	}()
+
+	// for UTF-8 BOM
+	_, err = file.WriteString("\xEF\xBB\xBF")
```

Comment:
Cannot understand here, can you add one link to UTF 8 BOM here?

Reply:
Write UTF-8 BOM mainly for unidentifiable Chinese code,
see https://stackoverflow.com/questions/2223882/whats-the-difference-between-utf-8-and-utf-8-without-bom